py/repl: Skip private variables when printing tab completion options.

andrewleech opened the pull request:
### Summary

The tab completion on repl is intended to not show any attributes starting with a single `_` character, however variables/functions in frozen modules are are currently shown.

This PR introduces a change to py/repl to skip private variables when printing tab completion options.

Originally included in https://github.com/micropython/micropython/pull/17011

### Testing

A unit test has added which covers this, currently based on the asyncio module which is generally frozen in. 
However looking again now maybe it would be better to add a suitable function / attribute to `./tests/frozen/frozentest.py` and test against that, assuming it should have the same behavior?

pi-anl committed:
py/repl: Skip private variables when printing tab completion options.

Any '_' variables/functions in frozen modules are currently printed, when
they shouldn't be.  That's due to underscore names possibly existing
between the start and end qstrs which are used to print the auto-complete
matches.  The underscore names should be skipped when iterating between the
two boundary qstrs.

The underscore attributes are removed from the extra coverage exp file
because tab completing "import <tab>" no longer lists modules beginning
with an underscore.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dpgeorge commented on py/repl.c:
```diff
@@ -218,7 +218,8 @@ static void print_completions(const mp_print_t *print,
         const char *d_str = (const char *)qstr_data(q, &d_len);
-        if (s_len <= d_len && strncmp(s_start, d_str, s_len) == 0) {
+        // check if item matches prefix and it's not a hidden attribute (starting with single underscore)
+        if (s_len <= d_len && strncmp(s_start, d_str, s_len) == 0 && (d_str[0] != '_' || (d_len > 1 && d_str[1] == '_'))) {
```
Wouldn't it be simpler and maybe smaller code to follow the pattern of `find_completions()` like this:
```c
        if (s_len == 0 && d_str[0] == '_') {
            continue;
        }                                        
        if (s_len <= d_len && strncmp(...) == 0) {
```
?

dpgeorge commented:
It's actually not easy to make a test for this that is robust because the bug relies on a certain ordering of qstrs in the frozen set of qstrs (it needs qstrs beginning with underscore to appear between the first and last items that match the tab completion).

So maybe it's not worth adding a test for?  Does the test you added fail prior to the fix in this PR?

andrewleech commented:
Curiosity - is the code size change different if it's all in one `if` statement?

dpgeorge commented:
CI needs fixing on this.

pi-anl committed:
tests/run-tests.py: Add support for ctrl keys in REPL tests.

This allows having {\xDD} in tests, which will be expanded to the given
hex character.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented:
@dpgeorge I've got a working test now I think should be reliable.

Added I minor change to the repl test support in run-tests.py to allow using ctrl-keys in these tests, can you think of any negative side effects of this?

dpgeorge commented:
> Added I minor change to the repl test support in run-tests.py to allow using ctrl-keys in these tests, can you think of any negative side effects of this?

I see it's only used for "special" tests, which limits its scope (that's good!).  So probably OK.

But, why does it need to run code using paste mode?  It seems like it should work just constructing that class normally?

dpgeorge commented on tests/cmdline/repl_autocomplete_underscore.py:
```diff
@@ -0,0 +1,34 @@
+
+# Test tab completion on the instance
+# The tab character after `obj.` below triggers the completion
+obj.	a	
```
Could it use the escape sequence for tab here, so it's clearer?  (Assuming the escape sequence is needed for paste mode, may as well use it here as well.)

andrewleech commented:
> But, why does it need to run code using paste mode? It seems like it should work just constructing that class normally?

When the class code it "typed" into repl via pty the auto-indent throws it off. I could flatten the class in the .py file, but then would still need backspace to de-dent in between each function.

dpgeorge commented:
> then would still need backspace to de-dent in between each function

The existing test `repl_autocomplete.py` has backspace in it as a raw byte.  You could follow that.

I'm not against adding an escape sequence parser, but if it's easy to work around that, might be better not to add such complexity?

dpgeorge commented:
> I didn't realise backspace and similar codes could just be added (with a hex editor etc) would you prefer that used for this / paste test, rather than the regex based parser?

No, I think what you added here is an improvement, much easier to see what's going on in the test with the escape sequences.

> Would you prefer a separate explicit paste test as well?

Yes, that would be nice, but not necessary.  Only if you have time.

> Taking a step back... the test I've added here seems to pass even without the fix in both `./run-tests.py cmdline/repl_autocomplete_underscore.py` and `./run-tests.py --via-mpy cmdline/repl_autocomplete_underscore.py` - is it only when frozen into the build it fails?

No, it's tricky.  It really depends on what qstrs are frozen and the order they appear in.  And the test behaviour may change if/when more scripts are frozen.  So I think we should just leave this PR as-is.

dpgeorge commented on tests/cmdline/repl_autocomplete_underscore.py:
```diff
@@ -0,0 +1,34 @@
+# Test REPL autocompletion filtering of underscore attributes
+# This test is designed to test behavior with frozen code when run with --via-mpy
```
Please remove this second line comment.  It's not valid.

dpgeorge commented:
CI is failing again...

pi-anl committed:
tests/cmdline: Add a test for REPL paste mode.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented:
> CI is failing again...

Fixed again; needed dos2unix on the exp files!.
I've added a repl paste test as well.
